homeassistant/components/freedompro/binary_sensor.py

```diff
@@ -0,0 +1,89 @@
+"""Support for Freedompro binary_sensor."""
+from homeassistant.components.binary_sensor import (
+    DEVICE_CLASS_MOTION,
+    DEVICE_CLASS_OCCUPANCY,
+    DEVICE_CLASS_OPENING,
+    DEVICE_CLASS_SMOKE,
+    BinarySensorEntity,
+)
+from homeassistant.const import CONF_API_KEY
+from homeassistant.core import callback
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up Freedompro binary_sensor."""
+    api_key = entry.data[CONF_API_KEY]
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        Device(hass, api_key, device, coordinator)
+        for device in coordinator.data
+        if device["type"] == "smokeSensor"
+        or device["type"] == "occupancySensor"
+        or device["type"] == "motionSensor"
+        or device["type"] == "contactSensor"
+    )
+
+
+class Device(CoordinatorEntity, BinarySensorEntity):
+    """Representation of an Freedompro binary_sensor."""
+
+    def __init__(self, hass, api_key, device, coordinator):
+        """Initialize the Freedompro binary_sensor."""
+        super().__init__(coordinator)
+        self._hass = hass
+        self._session = aiohttp_client.async_get_clientsession(self._hass)
+        self._api_key = api_key
+        self._attr_name = device["name"]
+        self._attr_unique_id = device["uid"]
+        self._type = device["type"]
+        self._characteristics = device["characteristics"]
+        self._attr_device_info = {
+            "name": self._attr_name,
+            "identifiers": {
+                (DOMAIN, self._attr_unique_id),
+            },
+            "model": self._type,
+            "manufacturer": "Freedompro",
+        }
```

Comment:
By doing this, code would be duplicated between light and binary_sensor.

Perhaps now is a good time to make a FreedomProDevice general class for all freedompro devices, then subclass it with the specific elements for light, sensor etc.

Also having multiple classes named Device could become confusing.

Reply:
hi, it might be a great idea, but since I have several pull requests at the same time, and the others are in the process of being approved, I would prefer to keep the parts divided and optimize them later.